Bugfix/zimbra 135 event logger shutdown fix

varunrisbud opened the pull request.

varunrisbud committed:
Fixed the eventLogger executorService shutdown issue

When EventLogger::shutdownEventLogger() was called the executorService
threads where being killed immediately. This was causing events in the
BatchingEventLogger not being committed to Solr.

To shutdown the executorService in a correct way a poison pill approach
is implemented. Poison pill events are placed in the event queue.
Whenever a thread encounters the poison pill it completes its execution.
Hence a grace full shutdown is possible.

Also to accommodate a use case to shutdown the executorService but not
to drain the event queue, the run method in EventNotifier is now using
eventQueue.poll(3, TimeUnit.SECONDS) method instead of eventQueue.take().
This does not blocks the EventNotifier till there is a event in the
eventQueue to know that the shutdown method was called and it needs to
complete execution. With the poll timeout set a thread will be blocked
for maximum of 3 seconds before it checks if shutdown is called.

varunrisbud committed:
Fixed an issue with closing a shared HttpClient resource

Whenever EventLogger::shutdownEventLogger is called, it also calls the
shutdown method on all the registered EventLogHandlers.
This was causing BatchingEventLogger to call the close method on
SolrEventCallback which in turn was calling the close method on
SolrCloudHelper. But SolrCloudHelper shared a same http client.
So closing the SolrCloudHelper by one BatchingEventLogger was causing
other BatchingEventLogger to error out and not able to send events to
Solr.

The HttpClient is created using SolrUtils::getCloudSolrClient(), which
in turn is using the
ZimbraHttpClientManager.getInstance().getInternalHttpClient().

varunrisbud committed:
Fixed a failing unit test

Came across this failing unit test while checking if anything is broke
due to changes to eventLogger. This unit test is not related to that fix
but it is a small change so fixing it.

The fix was needed due to changes in Solr schema done in ZIMBRA-95.

ilyusha commented on store/src/java/com/zimbra/cs/event/logger/EventLogger.java:
```diff
@@ -154,8 +156,14 @@ public void shutdownEventNotifierExecutor() {
         try {
+            if(drainQueueBeforeShutdown.get()) {
+                for (int i = 0; i < config.getNumThreads(); i++) {
+                    eventQueue.put(new Event("POISON_PILL_ACCOUNT_ID", Event.EventType.POISON_PILL, System.currentTimeMillis()));
```
This is a personal preference, so feel free to disregard - would a singleton poison pill make sense here? That way you don't need a `POISON_PILL_ACCOUNT_ID` string here... something like `eventQueue.put(Event.POISON_PILL)`?

ilyusha commented on store/src/java/com/zimbra/cs/event/logger/EventLogger.java:
```diff
@@ -197,29 +205,29 @@ private EventNotifier(Map<String, EventLogHandler.Factory> knownFactories, Map<S
+                Event event = null;
+                while (!shutdownExecutor.get()) {
+                    event = eventQueue.poll(3, TimeUnit.SECONDS);
+                    if (event != null && event.getEventType() != Event.EventType.POISON_PILL) {
```
Nitpick: import `EventType` directly so you don't need to keep repeating `Event.EventType.POISON_PILL`

varunrisbud committed:
Added POISON_PILL event as static variable for easy comparison

Create a static event called POISON_PILL of event type POISON_PILL.
It is used to add poison pills to the event queue. This way the same
instance of POISON_PILL is added to queue. So it is easier to write
comparison statements to see if event is of type POISON_PILL.

varunrisbud committed:
Added a static variable for poll timeout

Also added a few comments to make code more readable.

varunrisbud committed:
Poison pills are added to the event queue using offer with 3 sec timeout

In case of event queue being full we dont want to wait infinitely till
space is available. With the offer method we will wait for max of 3 sec
for each poison pill for space to become available. In worst case not
all EventNotifier threads will shutdown gracefully, but we decrease the
chance of waiting infinitely for shutdown to take place.

varunrisbud committed:
Removed the poison pill approach to shutdown the event logger

As pointed in the PR #485 the poison pill approach was not needed. We
opted to drain the event queue empty and then exit the EventNotifier
thread. With the current approach in case of empty queue we will need to
wait for 3 sec (EventNotifier.POLL_TIMEOUT seconds) to exit the
EventNotifier thread.
